File: Python DS & Algo - Runestone Academy/4. Basic Data Structures/C++/4.9_infix_to_postfix.cpp
#include<iostream>
#include <bits/stdc++.h>
#include<string>
using namespace std;

int precedence(char operand){
    int i;
    string operands = "^/*+-(";
    int precedences[] = {4, 3, 3, 2, 2, 1};
    for(i = 0; i < operands.length(); i++)
        if(operands[i] == operand)
            break;
    return precedences[i];
}

string infix_to_postfix(string expr){
    stack <char> op_stack;
    string postfix = "";
    for(int i = 0; i < expr.length(); i++){
        if(expr[i] == '(')
            op_stack.push(expr[i]);
        else if(isalpha(expr[i]))
            postfix += expr[i];
        else if(expr[i] == ')'){
            char top = op_stack.top();
            op_stack.pop();
            while(top != '(' && !op_stack.empty()){
                postfix += top;
                top = op_stack.top();
                op_stack.pop();
            }
        }
        else{
            while(!op_stack.empty() && precedence(op_stack.top()) >= precedence(expr[i])){
                postfix += op_stack.top();
                op_stack.pop();
            }
            op_stack.push(expr[i]);
        }
    }
    while(!op_stack.empty()){
        postfix += op_stack.top();
        op_stack.pop();
    }
    return postfix;
}

int main(void){
    cout << infix_to_postfix("A*B+C*D") << endl;                // A B * C D * +
    cout << infix_to_postfix("(A+B)*C-(D-E)*(F+G)") << endl;    // A B + C * D E - F G + * -
}

File: Python DS & Algo - Runestone Academy/4. Basic Data Structures/4.9_infix_to_postfix.py
from stack import Stack

def infix_to_postfix(expr):
    stack = Stack()
    postfix = ''
    precedence = {'*' : 3, '/' : 3, '+' : 2, '-' : 2, '(' : 1}
    for token in expr:
        if token.isalpha():
            postfix += token
        elif token == '(':
            stack.push(token)
        # Since the parentheses govern precedence, we have to make sure the
        # operations inside them are done first. This is why the operators are
        # popped immediately when a closing bracket is seen.
        elif token == ')':
            top_token = stack.pop()
            while top_token != '(' and not stack.isEmpty():
                postfix += top_token
                top_token = stack.pop()
        else:
            while not stack.isEmpty() and precedence[stack.peek()] >= precedence[token]:
                postfix += stack.pop()
            stack.push(token)
    while not stack.isEmpty():
        postfix += stack.pop()
    return postfix
print(infix_to_postfix("A*B+C*D"))
# A B * C D * +

print(infix_to_postfix("(A+B)*C-(D-E)*(F+G)"))
# A B + C * D E - F G + * -

print(infix_to_postfix('A+B*(C/D)+F*(G+H)')) # ABCD/*FGH+*++
